Check failure with python3.14

There were some changes to asyncio in python 3.14, that causes tests to fail:
```
============================= test session starts ==============================
platform linux -- Python 3.14.2, pytest-8.4.2, pluggy-1.6.0
rootdir: /build/python-eliot/src/eliot-1.17.5
plugins: hypothesis-6.150.0
collected 326 items

eliot/tests/test_api.py ..........                                       [  3%]
eliot/tests/test_coroutines.py F

=================================== FAILURES ===================================
____________ CoroutineTests.test_await_inherits_coroutine_contexts _____________

self = <eliot.tests.test_coroutines.CoroutineTests testMethod=test_await_inherits_coroutine_contexts>
logger = <eliot._output.MemoryLogger object at 0x7e69297d5c80>

    @capture_logging(None)
    def test_await_inherits_coroutine_contexts(self, logger):
        """
        awaited coroutines inherit the logging context.
        """
>       run_coroutines(calling_coro)

eliot/tests/test_coroutines.py:74: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
eliot/tests/test_coroutines.py:37: in run_coroutines
    loop = asyncio.get_event_loop()
           ^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <asyncio.unix_events._UnixDefaultEventLoopPolicy object at 0x7e692a23e7b0>

    def get_event_loop(self):
        """Get the event loop for the current context.
    
        Returns an instance of EventLoop or raises an exception.
        """
        if self._local._loop is None:
>           raise RuntimeError('There is no current event loop in thread %r.'
                               % threading.current_thread().name)
E           RuntimeError: There is no current event loop in thread 'MainThread'.

/usr/lib/python3.14/asyncio/events.py:715: RuntimeError
=========================== short test summary info ============================
FAILED eliot/tests/test_coroutines.py::CoroutineTests::test_await_inherits_coroutine_contexts
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
=================== 1 failed, 10 passed, 3 warnings in 0.61s ===================
```


related python issue: https://github.com/python/cpython/issues/126353

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -15,7 +15,7 @@ jobs:
 
     strategy:
       matrix:
-        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13", "pypy3.9", "pypy3.10"]
+        python-version: ["3.10", "3.11", "3.12", "3.13", "3.14", "pypy3.11"]
 
     steps:
       - uses: "actions/checkout@v3"

--- a/README.rst
+++ b/README.rst
@@ -25,7 +25,7 @@ Eliot supports a range of use cases and 3rd party libraries:
 
 Eliot is only used to generate your logs; you will might need tools like Logstash and ElasticSearch to aggregate and store logs if you are using multiple processes across multiple machines.
 
-Eliot supports Python 3.9-3.13, as well as PyPy3.
+Eliot supports Python 3.10-3.14, as well as PyPy 3.11.
 It is maintained by Itamar Turner-Trauring, and released under the Apache 2.0 License.
 
 * `Read the documentation <https://eliot.readthedocs.io>`_.

--- a/docs/source/news.rst
+++ b/docs/source/news.rst
@@ -1,6 +1,17 @@
 What's New
 ==========
 
+1.18.0
+^^^^^^
+
+Enhancements:
+
+* Official support for Python 3.14.
+
+Deprecations and removals:
+
+* Dropped support for Python 3.9.
+
 1.17.0
 ^^^^^^
 

--- a/eliot/_output.py
+++ b/eliot/_output.py
@@ -24,7 +24,6 @@ from .json import (
 )
 from ._validation import ValidationError
 
-
 # Action type for log messages due to a (hopefully temporarily) broken
 # destination.
 DESTINATION_FAILURE = "eliot:destination_failure"

--- a/eliot/_output.py
+++ b/eliot/_output.py
@@ -81,9 +81,6 @@ class Destinations(object):
         """
         message.update(self._globalFields)
         errors = []
-        is_destination_error_message = (
-            message.get("message_type", None) == DESTINATION_FAILURE
-        )
         for dest in self._destinations:
             try:
                 dest(message)
@@ -92,7 +89,7 @@ class Destinations(object):
                 # message, but rather continously, we will get a
                 # "eliot:destination_failure" log message logged, and so we
                 # want to ensure it doesn't do infinite recursion.
-                if not is_destination_error_message:
+                if message.get("message_type", None) != DESTINATION_FAILURE:
                     errors.append(e)
 
         for exception in errors:
@@ -139,7 +136,7 @@ class Destinations(object):
         if buffered_messages:
             # Re-deliver buffered messages:
             for message in buffered_messages:
-                self.send(message)
+                self.send(message, message.get("__eliot_logger__", None))
 
     def remove(self, destination):
         """

--- a/eliot/journald.py
+++ b/eliot/journald.py
@@ -12,11 +12,9 @@ from ._message import TASK_UUID_FIELD, MESSAGE_TYPE_FIELD
 from ._action import ACTION_TYPE_FIELD, ACTION_STATUS_FIELD, FAILED_STATUS
 
 _ffi = FFI()
-_ffi.cdef(
-    """
+_ffi.cdef("""
 int sd_journal_send(const char *format, ...);
-"""
-)
+""")
 try:
     try:
         _journald = _ffi.dlopen("libsystemd.so.0")

--- a/eliot/prettyprint.py
+++ b/eliot/prettyprint.py
@@ -19,7 +19,6 @@ from ._message import (
 )
 from ._action import ACTION_TYPE_FIELD, ACTION_STATUS_FIELD
 
-
 # Ensure binary stdin, since we expect specifically UTF-8 encoded
 # messages, not platform-encoding messages.
 stdin = stdin.buffer

--- a/eliot/tests/test_action.py
+++ b/eliot/tests/test_action.py
@@ -13,9 +13,6 @@ from hypothesis.strategies import integers, lists, just, text
 
 from pyrsistent import pvector, v
 
-import testtools
-from testtools.matchers import MatchesStructure
-
 from .._action import (
     Action,
     current_action,
@@ -961,10 +958,14 @@ class TaskLevelTests(TestCase):
         self.assertEqual(TaskLevel.to_string, TaskLevel.toString)
 
 
-class WrittenActionTests(testtools.TestCase):
+class WrittenActionTests(TestCase):
     """
     Tests for L{WrittenAction}.
     """
+
+    def assertWrittenActionContainsSubset(self, subset, action):
+        for key, value in subset.items():
+            self.assertEqual(getattr(action, key), value)
 
     @given(start_action_messages)
     def test_from_single_start_message(self, message):
@@ -974,9 +975,8 @@ class WrittenActionTests(testtools.TestCase):
         C{end_time}, and has a C{status} of C{STARTED_STATUS}.
         """
         action = WrittenAction.from_messages(message)
-        self.assertThat(
-            action,
-            MatchesStructure.byEquality(
+        self.assertWrittenActionContainsSubset(
+            dict(
                 status=STARTED_STATUS,
                 action_type=message.contents[ACTION_TYPE_FIELD],
                 task_uuid=message.task_uuid,
@@ -987,6 +987,7 @@ class WrittenActionTests(testtools.TestCase):
                 reason=None,
                 exception=None,
             ),
+            action,
         )
 
     @given(start_action_messages, message_dicts, integers(min_value=2))
@@ -1009,9 +1010,8 @@ class WrittenActionTests(testtools.TestCase):
             )
         )
         action = WrittenAction.from_messages(end_message=end_message)
-        self.assertThat(
-            action,
-            MatchesStructure.byEquality(
+        self.assertWrittenActionContainsSubset(
+            dict(
                 status=SUCCEEDED_STATUS,
                 action_type=end_message.contents[ACTION_TYPE_FIELD],
                 task_uuid=end_message.task_uuid,
@@ -1022,6 +1022,7 @@ class WrittenActionTests(testtools.TestCase):
                 reason=None,
                 exception=None,
             ),
+            action,
         )
 
     @given(message_dicts)
@@ -1033,9 +1034,8 @@ class WrittenActionTests(testtools.TestCase):
         """
         message = written_from_pmap(message_dict)
         action = WrittenAction.from_messages(children=[message])
-        self.assertThat(
-            action,
-            MatchesStructure.byEquality(
+        self.assertWrittenActionContainsSubset(
+            dict(
                 status=None,
                 action_type=None,
                 task_uuid=message.task_uuid,
@@ -1046,6 +1046,7 @@ class WrittenActionTests(testtools.TestCase):
                 reason=None,
                 exception=None,
             ),
+            action,
         )
 
     @given(start_action_messages, message_dicts, integers(min_value=2))
@@ -1165,9 +1166,8 @@ class WrittenActionTests(testtools.TestCase):
             )
         )
         action = WrittenAction.from_messages(start_message, end_message=end_message)
-        self.assertThat(
-            action,
-            MatchesStructure.byEquality(
+        self.assertWrittenActionContainsSubset(
+            dict(
                 action_type=start_message.contents[ACTION_TYPE_FIELD],
                 status=SUCCEEDED_STATUS,
                 task_uuid=start_message.task_uuid,
@@ -1178,6 +1178,7 @@ class WrittenActionTests(testtools.TestCase):
                 reason=None,
                 exception=None,
             ),
+            action,
         )
 
     @given(start_action_messages, message_dicts, text(), text(), integers(min_value=2))
@@ -1205,9 +1206,8 @@ class WrittenActionTests(testtools.TestCase):
             )
         )
         action = WrittenAction.from_messages(start_message, end_message=end_message)
-        self.assertThat(
-            action,
-            MatchesStructure.byEquality(
+        self.assertWrittenActionContainsSubset(
+            dict(
                 action_type=start_message.contents[ACTION_TYPE_FIELD],
                 status=FAILED_STATUS,
                 task_uuid=start_message.task_uuid,
@@ -1218,6 +1218,7 @@ class WrittenActionTests(testtools.TestCase):
                 reason=reason,
                 exception=exception,
             ),
+            action,
         )
 
     @given(start_action_messages, message_dicts, integers(min_value=2))

--- a/eliot/tests/test_action.py
+++ b/eliot/tests/test_action.py
@@ -647,7 +647,7 @@ class StartActionAndTaskTests(TestCase):
         action = startTask(action_type="sys:do", key="value")
         assertContainsFields(
             self,
-            messages[0],
+            messages[-1],
             {
                 "task_uuid": action._identification["task_uuid"],
                 "task_level": [1],
@@ -667,7 +667,7 @@ class StartActionAndTaskTests(TestCase):
         action = start_action(action_type="sys:do", key="value")
         assertContainsFields(
             self,
-            messages[0],
+            messages[-1],
             {
                 "task_uuid": action._identification["task_uuid"],
                 "task_level": [1],
@@ -1487,15 +1487,22 @@ class PreserveContextTests(TestCase):
         Message.log(message_type="child")
         return x + y
 
-    def test_no_context(self):
+    @capture_logging(None)
+    def test_no_context(self, logger):
         """
         If C{preserve_context} is run outside an action context it just
         returns the same function.
         """
-        wrapped = preserve_context(self.add)
+        # Make sure no leaks from previous tests:
+        assert current_action() is None
+
+        add = self.add
+        wrapped = preserve_context(add)
+        self.assertIs(add, wrapped)
         self.assertEqual(wrapped(2, 3), 5)
 
-    def test_with_context_calls_underlying(self):
+    @capture_logging(None)
+    def test_with_context_calls_underlying(self, logger):
         """
         If run inside an Eliot context, the result of C{preserve_context} is
         the result of calling the underlying function.
@@ -1527,7 +1534,8 @@ class PreserveContextTests(TestCase):
             ("parent", "eliot:remote_task", "child"),
         )
 
-    def test_callable_only_once(self):
+    @capture_logging(None)
+    def test_callable_only_once(self, logger):
         """
         The result of C{preserve_context} can only be called once.
         """

--- a/eliot/tests/test_coroutines.py
+++ b/eliot/tests/test_coroutines.py
@@ -34,7 +34,8 @@ def run_coroutines(*async_functions):
     """
     Run a coroutine until it finishes.
     """
-    loop = asyncio.get_event_loop()
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
     futures = [asyncio.ensure_future(f()) for f in async_functions]
 
     async def wait_for_futures():

--- a/eliot/tests/test_message.py
+++ b/eliot/tests/test_message.py
@@ -185,6 +185,7 @@ class DeprecatedMessageTests(TestCase):
         Message.new(key=2).write(logger)
         Message.new(key=3).write(logger)
 
+        self.assertEqual(len(logger.messages), 2, logger.messages)
         message1, message2 = logger.messages
         self.assertEqual(
             (
@@ -287,9 +288,9 @@ class LogMessageTests(TestCase):
         add_destinations(messages.append)
         self.addCleanup(remove_destination, messages.append)
         log_message(message_type="hello", some_key=1234)
-        self.assertEqual(messages[0]["some_key"], 1234)
-        self.assertEqual(messages[0]["message_type"], "hello")
-        self.assertTrue(time.time() - messages[0]["timestamp"] < 0.1)
+        self.assertEqual(messages[-1]["some_key"], 1234)
+        self.assertEqual(messages[-1]["message_type"], "hello")
+        self.assertTrue(time.time() - messages[-1]["timestamp"] < 0.1)
 
     def test_implicitAction(self):
         """

--- a/eliot/tests/test_output.py
+++ b/eliot/tests/test_output.py
@@ -367,7 +367,7 @@ class DestinationsTests(TestCase):
         dest3 = []
         destinations.add(dest.append, dest2.append)
         destinations.add(dest3.append)
-        destinations.send(message)
+        destinations.send(message, logger=MemoryLogger())
         self.assertEqual(dest, [message])
         self.assertEqual(dest2, [message])
         self.assertEqual(dest3, [message])
@@ -386,7 +386,7 @@ class DestinationsTests(TestCase):
         destinations.add(dest3.append)
 
         message = {"hello": 123}
-        destinations.send(message)
+        destinations.send(message, logger=MemoryLogger())
         self.assertIn(message, dest)
         self.assertIn(message, dest3)
 
@@ -401,9 +401,9 @@ class DestinationsTests(TestCase):
 
         msg1 = {"hello": 123}
         msg2 = {"world": 456}
-        destinations.send(msg1)
+        destinations.send(msg1, logger=MemoryLogger())
         self.assertNotIn(msg1, dest)
-        destinations.send(msg2)
+        destinations.send(msg2, logger=MemoryLogger())
         self.assertIn(msg2, dest)
 
     def test_remove(self):
@@ -416,7 +416,7 @@ class DestinationsTests(TestCase):
         dest = []
         destinations.add(dest.append)
         destinations.remove(dest.append)
-        destinations.send(message)
+        destinations.send(message, logger=MemoryLogger())
         self.assertEqual(dest, [])
 
     def test_removeNonExistent(self):
@@ -436,7 +436,7 @@ class DestinationsTests(TestCase):
         dest = []
         destinations.add(dest.append)
         destinations.addGlobalFields(x=123, y="hello")
-        destinations.send({"z": 456})
+        destinations.send({"z": 456}, MemoryLogger())
         self.assertEqual(dest, [{"x": 123, "y": "hello", "z": 456}])
 
     def test_addGlobalFieldsCumulative(self):
@@ -449,7 +449,7 @@ class DestinationsTests(TestCase):
         destinations.add(dest.append)
         destinations.addGlobalFields(x=123, y="hello")
         destinations.addGlobalFields(x=456, z=456)
-        destinations.send({"msg": "X"})
+        destinations.send({"msg": "X"}, MemoryLogger())
         self.assertEqual(dest, [{"x": 456, "y": "hello", "z": 456, "msg": "X"}])
 
     def test_buffering(self):
@@ -460,7 +460,7 @@ class DestinationsTests(TestCase):
         destinations = Destinations()
         messages = [{"k": i} for i in range(1050)]
         for m in messages:
-            destinations.send(m)
+            destinations.send(m, MemoryLogger())
         dest, dest2 = [], []
         destinations.add(dest.append, dest2.append)
         self.assertEqual((dest, dest2), (messages[-1000:], messages[-1000:]))
@@ -472,12 +472,12 @@ class DestinationsTests(TestCase):
         destinations = Destinations()
         message = {"m": 1}
         message2 = {"m": 2}
-        destinations.send(message)
+        destinations.send(message, MemoryLogger())
         dest = []
         dest2 = []
         destinations.add(dest.append)
         destinations.add(dest2.append)
-        destinations.send(message2)
+        destinations.send(message2, MemoryLogger())
         self.assertEqual((dest, dest2), ([message, message2], [message2]))
 
     def test_global_fields_buffering(self):
@@ -486,7 +486,7 @@ class DestinationsTests(TestCase):
         """
         destinations = Destinations()
         message = {"m": 1}
-        destinations.send(message)
+        destinations.send(message, MemoryLogger())
         destinations.addGlobalFields(k=123)
         dest = []
         destinations.add(dest.append)

--- a/eliot/tests/test_testing.py
+++ b/eliot/tests/test_testing.py
@@ -695,7 +695,7 @@ class CaptureLoggingTests(ValidateLoggingTestsMixin, TestCase):
         add_destination(messages.append)
         self.addCleanup(remove_destination, messages.append)
         Message.log(some_key=1234)
-        self.assertEqual(messages[0]["some_key"], 1234)
+        self.assertEqual(messages[0]["some_key"], 1234, messages)
 
     def test_global_cleanup_exception(self):
         """

--- a/eliot/tests/test_validation.py
+++ b/eliot/tests/test_validation.py
@@ -541,9 +541,9 @@ class MessageTypeTests(TestCase):
         self.addCleanup(remove_destination, messages.append)
         message_type = self.messageType()
         message_type.log(key=1234, value=3)
-        self.assertEqual(messages[0]["key"], 1234)
-        self.assertEqual(messages[0]["value"], 3)
-        self.assertEqual(messages[0]["message_type"], message_type.message_type)
+        self.assertEqual(messages[-1]["key"], 1234, messages[0])
+        self.assertEqual(messages[-1]["value"], 3)
+        self.assertEqual(messages[-1]["message_type"], message_type.message_type)
 
     def test_description(self):
         """

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,0 +1,3 @@
+[build-system]
+requires = ["setuptools", "setuptools-scm", "versioneer"]
+build-backend = "setuptools.build_meta"

--- a/setup.py
+++ b/setup.py
@@ -18,11 +18,11 @@ setup(
         "Operating System :: OS Independent",
         "Programming Language :: Python",
         "Programming Language :: Python :: 3",
-        "Programming Language :: Python :: 3.9",
         "Programming Language :: Python :: 3.10",
         "Programming Language :: Python :: 3.11",
         "Programming Language :: Python :: 3.12",
         "Programming Language :: Python :: 3.13",
+        "Programming Language :: Python :: 3.14",
         "Programming Language :: Python :: Implementation :: CPython",
         "Programming Language :: Python :: Implementation :: PyPy",
         "Topic :: System :: Logging",
@@ -31,7 +31,7 @@ setup(
     version=versioneer.get_version(),
     cmdclass=versioneer.get_cmdclass(),
     description="Logging library that tells you why it happened",
-    python_requires=">=3.9.0",
+    python_requires=">=3.10.0",
     install_requires=[
         # Internal code documentation:
         "zope.interface",

--- a/setup.py
+++ b/setup.py
@@ -50,8 +50,6 @@ setup(
         "test": [
             # Bug-seeking missile:
             "hypothesis >= 1.14.0",
-            # Tasteful testing for Python:
-            "testtools",
             "pytest",
             "pytest-xdist",
         ],

--- a/tox.ini
+++ b/tox.ini
@@ -16,7 +16,7 @@ python =
 commands =
     pip install .[test]
     pip list
-    {envpython} -m pytest -n auto
+    {envpython} -m pytest
 
 ## No Twisted ##
 [testenv:py313-numpy]

--- a/tox.ini
+++ b/tox.ini
@@ -1,28 +1,26 @@
 [tox]
-envlist = py39-numpy, pypy39, pypy310, py38, py39, py310, py311, py312, py313,
-          py39-twisted-latest,
+envlist = py313-numpy, pypy311, py310, py311, py312, py313, py314,
+          py313-twisted-latest,
           linters3, sphinx
 
 [gh-actions]
 python =
-       3.9: py39, linters3, sphinx, py39-numpy, py39-twisted-latest
        3.10: py310
        3.11: py311
        3.12: py312
-       3.13: py313
-       pypy3.9: pypy39
-       pypy3.10: pypy310
+       3.13: py313, linters3, sphinx, py313-numpy, py313-twisted-latest
+       3.14: py314
+       pypy3.11: pypy311
 
 [testenv]
 commands =
-    {envpython} setup.py --version
     pip install .[test]
     pip list
     {envpython} -m pytest -n auto
 
 ## No Twisted ##
-[testenv:py39-numpy]
-basepython = python3.9
+[testenv:py313-numpy]
+basepython = python3.13
 deps = numpy
        dask[bag]
 
@@ -38,13 +36,8 @@ deps = cffi
        polars
        pydantic
 
-[testenv:py39]
-basepython = python3.9
-deps = cffi
-       pyinstaller
-
-[testenv:py39-twisted-latest]
-basepython = python3.9
+[testenv:py313-twisted-latest]
+basepython = python3.13
 deps = Twisted
 
 [testenv:pypy3]
@@ -56,14 +49,14 @@ deps =
 deps = pyflakes
        black
        flake8
-basepython = python3.9
+basepython = python3.13
 commands =
          pyflakes eliot
-         black --check --target-version=py39 eliot
+         black --check --target-version=py310 eliot
 
 [testenv:sphinx]
 deps = sphinx
        sphinx_rtd_theme
-basepython = python3.9
+basepython = python3.13
 commands =
     sphinx-build -b html docs/source/ docs/_build/html

